samza-core/src/main/java/org/apache/samza/control/WatermarkManager.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.control;
+
+import com.google.common.collect.Multimap;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import org.apache.samza.message.WatermarkMessage;
+import org.apache.samza.system.IncomingMessageEnvelope;
+import org.apache.samza.system.StreamMetadataCache;
+import org.apache.samza.system.SystemStream;
+import org.apache.samza.system.SystemStreamPartition;
+import org.apache.samza.task.MessageCollector;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This class manages watermarks. It aggregates the watermark control messages from the upstage tasks
+ * for each SSP into an envelope of {@link Watermark}, and provide a dispatcher to propagate it to downstream.
+ */
+public class WatermarkManager {
+  private static final Logger log = LoggerFactory.getLogger(WatermarkManager.class);
+  public static final long TIME_NOT_EXIST = -1;
+
+  private final String taskName;
+  private final Map<SystemStreamPartition, WatermarkState> watermarkStates;
+  private final Map<SystemStream, Long> watermarkPerStream;
+  private final StreamMetadataCache metadataCache;
+  // mapping from input stream to its downstream tasks
+  private final Multimap<SystemStream, String> inputToTasks;
+  private final MessageCollector collector;
+
+  // mapping from output stream to its upstream task count
+  private Map<SystemStream, Integer> upstreamTaskCounts;
```

Comment:
As Yi mentioned below, it would be great to eliminate init and move to final to the greatest degree possible.